homeassistant/components/mqtt/mixins.py

```diff
@@ -604,6 +605,7 @@ async def cleanup_device_registry(
     entity_registry = er.async_get(hass)
     if (
         device_id
+        and device_id not in device_registry.deleted_devices
```

Comment:
This is most likely a bug. I'll open a PR to take this on out.

Reply:
Seems it is a side effect with the new code, it does not seem to happen when using in a manual cleanup with 2 single discovery topics (test added)

Reply:
It seems to be race with device based discovery. The condition will avoid removing a device that already has been cleaned up.